src/meta/src/hummock/compaction/selector/level_selector.rs

```diff
@@ -239,53 +239,62 @@ impl DynamicLevelSelectorCore {
-                let base_level_sst_count =
-                    levels.get_level(ctx.base_level).table_infos.len() as u64;
-
-                // size limit
-                let non_overlapping_size_score = total_size * SCORE_BASE
-                    / std::cmp::max(self.config.max_bytes_for_level_base, base_level_size);
-                // level count limit
-                let non_overlapping_level_count = levels
-                    .l0
-                    .as_ref()
-                    .unwrap()
-                    .sub_levels
-                    .iter()
-                    .filter(|level| level.level_type() == LevelType::Nonoverlapping)
-                    .count() as u64;
-                let non_overlapping_level_score = non_overlapping_level_count * SCORE_BASE
-                    / std::cmp::max(
-                        base_level_sst_count / 16,
-                        self.config.level0_sub_level_compact_level_count as u64,
-                    );
-
-                std::cmp::max(non_overlapping_size_score, non_overlapping_level_score)
-            };
+            let total_size = levels
+                .l0
+                .as_ref()
+                .unwrap()
+                .sub_levels
+                .iter()
+                .filter(|level| {
+                    level.vnode_partition_count == levels.vnode_partition_count
+                        && level.level_type() == LevelType::Nonoverlapping
+                })
+                .map(|level| level.total_file_size)
+                .sum::<u64>()
+                - handlers[0].get_pending_output_file_size(ctx.base_level as u32);
+            let base_level_size = levels.get_level(ctx.base_level).total_file_size;
+            let base_level_sst_count = levels.get_level(ctx.base_level).table_infos.len() as u64;
+
+            // size limit
+            let non_overlapping_size_score = total_size * SCORE_BASE
+                / std::cmp::max(self.config.max_bytes_for_level_base, base_level_size);
+            // level count limit
+            let non_overlapping_level_count = levels
+                .l0
+                .as_ref()
+                .unwrap()
+                .sub_levels
+                .iter()
+                .filter(|level| level.level_type() == LevelType::Nonoverlapping)
+                .count() as u64;
+            let non_overlapping_level_score = non_overlapping_level_count * SCORE_BASE
+                / std::cmp::max(
+                    base_level_sst_count / 16,
+                    self.config.level0_sub_level_compact_level_count as u64,
+                );
+
+            let non_overlapping_score =
+                std::cmp::max(non_overlapping_size_score, non_overlapping_level_score);
 
             // Reduce the level num of l0 non-overlapping sub_level
-            ctx.score_levels.push({
-                PickerInfo {
+            if non_overlapping_size_score > SCORE_BASE {
```

Comment:
Could you explain the idea behind non_overlapping_size_score > SCORE_BASE and non_overlapping_level_score > SCORE_BASE? Why ToBase is stricter about size_score, while Intra is stricter about level_score?

Reply:
because if there are not enough split-vnode file in level0, the selector will pick a large task to base level and it will case more write-amplification and also slowdown compaction